perl/cpan/private/carton.bzl

```
@@ -0,0 +1,199 @@
+"""Bazel tools for interfacing with Carton"""
+
+_HUB_BUILD_FILE = """\
+load("@rules_perl//perl:perl_library.bzl", "perl_library")
+
+package(default_visibility = ["//visibility:public"])
+
+DEPENDENCIES = {dependencies}
+
+perl_library(
+    name = "{name}",
+    deps = [
+        "@{name}__" + dep
+        for dep in DEPENDENCIES
+    ],
+    includes = [],
+)
+
+[
+    alias(
+        name = dep,
+        actual = "@{name}__" + dep,
+    )
+    for dep in DEPENDENCIES
+]
+"""
+
+def _perl_cpan_hub_impl(repository_ctx):
+    repository_ctx.file("WORKSPACE.bazel", """workspace(name = "{}")""".format(
+        repository_ctx.name,
+    ))
+
+    repository_ctx.file("BUILD.bazel", _HUB_BUILD_FILE.format(
+        name = repository_ctx.attr.hub_name,
+        dependencies = json.encode_indent(repository_ctx.attr.modules, indent = " " * 4),
+    ))
+
+perl_cpan_hub = repository_rule(
+    doc = "A hub repository that exposes all CPAN dependencies from a `cpanfile`.",
+    implementation = _perl_cpan_hub_impl,
+    attrs = {
+        "hub_name": attr.string(
+            doc = "The name of the hub.",
+            mandatory = True,
+        ),
+        "modules": attr.string_list(
+            doc = "Dependencies to add to the hub.",
+            mandatory = True,
+        ),
+    },
+)
+
+_CPAN_MODULE_BUILD_FILE = """\
+load("@rules_perl//perl:perl_library.bzl", "perl_library")
```

Why use the private interface here? I think I would do
```suggestion
load("@rules_perl//perl:perl.bzl", "perl_library")
```
So this doesn't need to change if perl_library changes and also to not confuse users who may read this BUILD when debugging

---

fair, maybe the impl should be moved to a more private place 😅